resources/user.go

```diff
@@ -0,0 +1,212 @@
+// Mgmt
+// Copyright (C) 2013-2017+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	errwrap "github.com/pkg/errors"
+	"log"
+	"os/user"
+)
+
+func init() {
+	RegisterResource("user", func() Res { return &UserRes{} })
+}
+
+//TODO
+type UserRes struct {
+	BaseRes `yaml:",inline"`
+	Gid     string `yaml:"gid"`
+	HomeDir string `yaml:"homedir"`
+	State   string `yaml:"state"` // state: exists, absent
+	Uid     string `yaml:"uid"`
+	exec    *ExecRes
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *UserRes) Default() Res {
+	return &UserRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *UserRes) Validate() error {
+	if obj.State != "exists" && obj.State != "absent" {
+		return fmt.Errorf("State must be 'exists' or 'absent'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *UserRes) Init() error {
+	obj.exec = &ExecRes{}
+	obj.exec.Init()
```

Comment:
Great, but you forgot to check return value of exec's Init()

Reply:
I initially removed the check because it segfaults when I run it:. What am I doing wrong?

    if err := obj.exec.Init(); err != nil {
        return err
    }

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x482a2d]

goroutine 141 [running]:
sync.(*Mutex).Lock(0x0)
        /usr/lib/go-1.8/src/sync/mutex.go:48 +0x2d
github.com/purpleidea/mgmt/resources.(*BaseRes).ProcessExit(0xc4201d4000)
        /home/jon/gopath/src/github.com/purpleidea/mgmt/resources/resources.go:439 +0x32
github.com/purpleidea/mgmt/resources.(*BaseRes).Worker(0xc4201d4000, 0x0, 0x0)
        /home/jon/gopath/src/github.com/purpleidea/mgmt/resources/actions.go:515 +0xb1
github.com/purpleidea/mgmt/resources.(*MGraph).Start.func2(0xc420284780, 0x7fa389027350, 0xc4201d4000)
        /home/jon/gopath/src/github.com/purpleidea/mgmt/resources/mgraph.go:163 +0x17a
created by github.com/purpleidea/mgmt/resources.(*MGraph).Start
        /home/jon/gopath/src/github.com/purpleidea/mgmt/resources/mgraph.go:168 +0x716
jon@jon-ThinkPad-E470:~/mgmt (feat/user_res)$